File: app/src/main/java/social/entourage/android/base/EntouragePagination.java
package social.entourage.android.base;

import java.util.Date;

import social.entourage.android.Constants;

/**
 * Created by mihaiionescu on 10/05/16.
 */
public class EntouragePagination {

    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------
    public int page = 1;
    public int itemsPerPage;
    protected Date beforeDate = new Date();
    protected Date newestDate = null;
    public boolean isLoading = false;
    public boolean isRefreshing = false;
    public boolean newItemsAvailable = false;
    public boolean nextPageAvailable = false;

    // ----------------------------------
    // CONSTRUCTORS
    // ---------------------------------

    public EntouragePagination() {
        this.itemsPerPage = Constants.ITEMS_PER_PAGE;
    }

    public EntouragePagination(int itemsPerPage) {
        this.itemsPerPage = itemsPerPage;
    }

    // ----------------------------------
    // GETTERS AND SETTERS
    // ---------------------------------

    public Date getBeforeDate() {
        return (isRefreshing ? new Date() : beforeDate);
    }

    public void setBeforeDate(final Date beforeDate) {
        this.beforeDate = beforeDate;
    }


    // ----------------------------------
    // METHODS
    // ---------------------------------

    public void loadedItems(int loadedItems) {
        nextPageAvailable = false;
        if (!isRefreshing && loadedItems >= itemsPerPage) {
            page++;
            nextPageAvailable = true;
        }
        isLoading = false;
    }

    public void loadedItems(Date newestDate, Date oldestDate) {
        if (newestDate == null) {
            newItemsAvailable = false;
        }
        else {
            if (this.newestDate == null) {
                this.newestDate = newestDate;
            }
            if (!isRefreshing) {
                beforeDate = oldestDate;
            }
            else {
                if (this.newestDate.before(newestDate)) {
                    this.newestDate = newestDate;
                    newItemsAvailable = true;
                }
            }
        }
        isLoading = false;
    }

}

File: app/src/main/java/social/entourage/android/base/EntouragePagination.kt
package social.entourage.android.base

import social.entourage.android.Constants
import java.util.*

/**
 * Created by mihaiionescu on 10/05/16.
 */
open class EntouragePagination {
    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------
    var page = 1
    @JvmField
    var itemsPerPage: Int
    var beforeDate:Date
        get() = if (isRefreshing) Date() else field
        //set(newBeforeDate) { this.beforeDate = newBeforeDate}
    protected var newestDate: Date? = null
    @JvmField
    var isLoading = false
    @JvmField
    var isRefreshing = false
    var newItemsAvailable = false
    var nextPageAvailable = false

    // ----------------------------------
    // CONSTRUCTORS
    // ---------------------------------
    constructor() {
        itemsPerPage = Constants.ITEMS_PER_PAGE
    }

    constructor(itemsPerPage: Int) {
        this.itemsPerPage = itemsPerPage
    }

    open fun reset() {
        page = 1
        itemsPerPage = Constants.ITEMS_PER_PAGE
        beforeDate = Date()
        newestDate = null
        isLoading = false
        isRefreshing = false
        newItemsAvailable = false
        nextPageAvailable = false
    }

    // ----------------------------------
    // METHODS
    // ---------------------------------
    fun loadedItems(loadedItems: Int) {
        nextPageAvailable = false
        if (!isRefreshing && loadedItems >= itemsPerPage) {
            page++
            nextPageAvailable = true
        }
        isLoading = false
    }

    fun loadedItems() {
        newItemsAvailable = false
        isLoading = false
    }

    fun loadedItems(newestDate: Date, oldestDate: Date) {
        if (this.newestDate == null) {
            this.newestDate = newestDate
        }
        if (!isRefreshing) {
            beforeDate = oldestDate
        } else {
            if (this.newestDate?.before(newestDate)==true) {
                this.newestDate = newestDate
                newItemsAvailable = true
            }
        }
        isLoading = false
    }

    init {
        beforeDate = Date()
    }
}